request/graphql/schema/collection.go

```diff
@@ -159,15 +132,211 @@ func fromAstDefinition(
+	for i := 1; i < len(name); i++ {
+		c := name[i]
+		if (c < 'a' || c > 'z') && (c < 'A' || c > 'Z') && (c < '0' || c > '9') && c != '_' {
+			return false
+		}
+	}
+	return true
+}
+
+func fieldIndexFromAST(field *ast.FieldDefinition, directive *ast.Directive) (client.IndexDescription, error) {
+	desc := client.IndexDescription{
+		Fields: []client.IndexedFieldDescription{
+			{
+				Name:      field.Name.Value,
+				Direction: client.Ascending,
+			},
+		},
+	}
+	for _, arg := range directive.Arguments {
+		switch arg.Name.Value {
+		case indexDirectivePropName:
+			nameVal, ok := arg.Value.(*ast.StringValue)
+			if !ok {
+				return client.IndexDescription{}, ErrIndexWithInvalidArg
+			}
+			desc.Name = nameVal.Value
+			if !isValidIndexName(desc.Name) {
+				return client.IndexDescription{}, ErrIndexWithInvalidArg
+			}
+		case indexDirectivePropUnique:
+			boolVal, ok := arg.Value.(*ast.BooleanValue)
+			if !ok {
+				return client.IndexDescription{}, ErrIndexWithInvalidArg
+			}
+			desc.Unique = boolVal.Value
+		default:
+			return client.IndexDescription{}, ErrIndexWithUnknownArg
+		}
+	}
+	return desc, nil
+}
+
+func indexFromAST(directive *ast.Directive) (client.IndexDescription, error) {
+	desc := client.IndexDescription{}
+	var directions *ast.ListValue
+	for _, arg := range directive.Arguments {
+		switch arg.Name.Value {
+		case indexDirectivePropName:
+			nameVal, ok := arg.Value.(*ast.StringValue)
+			if !ok {
+				return client.IndexDescription{}, ErrIndexWithInvalidArg
+			}
+			desc.Name = nameVal.Value
+			if !isValidIndexName(desc.Name) {
+				return client.IndexDescription{}, ErrIndexWithInvalidArg
+			}
+		case indexDirectivePropFields:
+			fieldsVal, ok := arg.Value.(*ast.ListValue)
+			if !ok {
+				return client.IndexDescription{}, ErrIndexWithInvalidArg
+			}
+			for _, field := range fieldsVal.Values {
+				fieldVal, ok := field.(*ast.StringValue)
+				if !ok {
+					return client.IndexDescription{}, ErrIndexWithInvalidArg
+				}
+				desc.Fields = append(desc.Fields, client.IndexedFieldDescription{
+					Name: fieldVal.Value,
+				})
+			}
+		case indexDirectivePropDirections:
+			var ok bool
+			directions, ok = arg.Value.(*ast.ListValue)
+			if !ok {
+				return client.IndexDescription{}, ErrIndexWithInvalidArg
+			}
+		case indexDirectivePropUnique:
+			boolVal, ok := arg.Value.(*ast.BooleanValue)
+			if !ok {
+				return client.IndexDescription{}, ErrIndexWithInvalidArg
+			}
+			desc.Unique = boolVal.Value
+		default:
+			return client.IndexDescription{}, ErrIndexWithUnknownArg
+		}
+	}
+	if len(desc.Fields) == 0 {
+		return client.IndexDescription{}, ErrIndexMissingFields
+	}
+	if directions != nil {
+		if len(directions.Values) != len(desc.Fields) {
+			return client.IndexDescription{}, ErrIndexWithInvalidArg
+		}
+		for i := range desc.Fields {
+			dirVal, ok := directions.Values[i].(*ast.EnumValue)
+			if !ok {
+				return client.IndexDescription{}, ErrIndexWithInvalidArg
+			}
+			if dirVal.Value == "ASC" {
```

Comment:
todo: `"ASC"` and `"DESC"` should really be using constants declared in the `client/requests` package I think, instead of duplicating the underlying strings.  It makes it much easier to identify the related components if nothing else.